lib/auth/join_github.go

```diff
@@ -39,7 +39,14 @@ func (a *Server) checkGitHubJoinRequest(ctx context.Context, req *types.Register
 		return trace.Wrap(err)
 	}
 
-	claims, err := a.ghaIDTokenValidator.Validate(ctx, req.IDToken)
+	token, ok := pt.(*types.ProvisionTokenV2)
```

Comment:
Yeah, here we see a "type assertion" where "ok" will be a boolean that states whether or not that concrete type has been able to be extracted from the interface.

Reply:
not a complaint about your PR here, but in general I wish we were more consistent about using either interfaces or plain struct types. The ProvisionToken interface is just getting in the way now that it is asserted to ProvisionTokenV2 in multiple places

Reply:
Yup @nklaassen - this is an outstanding concern of mine around the joining code.

I can see some argument for adding the following to the ProvisionToken interface:

```go
type ProvisionToken interface {
    GitHubConfig() *types.ProvisionTokenSpecV2GitHub
}
```

Whilst this gets rid of the assertion here, it seems antithetical to the purpose of these interfaces, which seems to be broadly about "compatibility": allowing us to introduce a new ProvisionToken version and for this new version of ProvisionToken to be used in place of the old version. In order to follow this vision of "compatibility" we'd need to introduce an interface for the GitHub config itself, something like:

```go
type ProvisionToken interface {
    GitHubConfig() GitHubConfig
}

type GitHubConfig {
    GHESHost() string
}
```

All of this pursuit of "compatibility" is very noble, but the layers and layers of interfaces definitely makes the code harder to navigate and far more cumbersome for us to edit. I generally find the definition of these interfaces alongside with the types to be extremely "un-go", I prefer to define interfaces where they are consumed (although this isn't an extremely convincing argument for us to not do it).

If, as an org, we were actually regularly making use of new Resource versions in order to make breaking changes to the structure of resources, then this would be an easy decision to make. However, my current understanding is that this is not a strategy we tend to pursue. The current format of the Teleport KV store means that introducing a new resource version completely breaks support for downgrading an auth server once a resource of the new type has been created.

I could well be making a mountain out of a molehill here, but, it would certainly be my preference that if we aren't going to be leveraging the interfaces regularly to support multiple versions of a resource, that we'd just use the concrete types (which would certainly be simpler !)